{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Creating TCGA cohorts  (part 1)\n",
    "\n",
    "This notebook will show you how to create a TCGA cohort using the publicly available TCGA BigQuery tables that the [ISB-CGC](http://isb-cgc.org) project has produced based on the open-access [TCGA](http://cancergenome.nih.gov/) data available at the [Data Portal](https://tcga-data.nci.nih.gov/tcga/).  You will need to have access to a Google Cloud Platform (GCP) project in order to use BigQuery.  If you don't already have one, you can sign up for a [free-trial](https://cloud.google.com/free-trial/) or contact [us](mailto://info@isb-cgc.org) and become part of the community evaluation phase of our [Cancer Genomics Cloud pilot](https://cbiit.nci.nih.gov/ncip/nci-cancer-genomics-cloud-pilots).\n",
    "\n",
    "We are not attempting to provide a thorough BigQuery or IPython tutorial here, as a wealth of such information already exists.  Here are some links to some resources that you might find useful: \n",
    "* [BigQuery](https://cloud.google.com/bigquery/what-is-bigquery), \n",
    "* the BigQuery [web UI](https://bigquery.cloud.google.com/) where you can run queries interactively, \n",
    "* [IPython](http://ipython.org/) (now known as [Jupyter](http://jupyter.org/)), and \n",
    "* [Cloud Datalab](https://cloud.google.com/datalab/) the recently announced interactive cloud-based platform that this notebook is being developed on. \n",
    "\n",
    "There are also many tutorials and samples available on github (see, in particular, the [datalab](https://github.com/GoogleCloudPlatform/datalab) repo and the [Google Genomics](  https://github.com/googlegenomics) project).\n",
    "\n",
    "OK then, let's get started!  In order to work with BigQuery, the first thing you need to do is import the bigquery module:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: google-cloud-bigquery==0.30 in /usr/local/lib/python2.7/dist-packages\n",
      "Requirement already satisfied: google-auth>=1.0.0 in /usr/local/lib/python2.7/dist-packages (from google-cloud-bigquery==0.30)\n",
      "Requirement already satisfied: google-cloud-core<0.29dev,>=0.28.0 in /usr/local/lib/python2.7/dist-packages (from google-cloud-bigquery==0.30)\n",
      "Requirement already satisfied: requests>=2.18.0 in /usr/local/lib/python2.7/dist-packages (from google-cloud-bigquery==0.30)\n",
      "Requirement already satisfied: google-resumable-media>=0.2.1 in /usr/local/lib/python2.7/dist-packages (from google-cloud-bigquery==0.30)\n",
      "Requirement already satisfied: google-api-core<0.2.0dev,>=0.1.1 in /usr/local/lib/python2.7/dist-packages (from google-cloud-bigquery==0.30)\n",
      "Requirement already satisfied: rsa>=3.1.4 in /usr/local/lib/python2.7/dist-packages (from google-auth>=1.0.0->google-cloud-bigquery==0.30)\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in /usr/local/lib/python2.7/dist-packages (from google-auth>=1.0.0->google-cloud-bigquery==0.30)\n",
      "Requirement already satisfied: six>=1.9.0 in /usr/local/lib/python2.7/dist-packages (from google-auth>=1.0.0->google-cloud-bigquery==0.30)\n",
      "Requirement already satisfied: cachetools>=2.0.0 in /usr/local/lib/python2.7/dist-packages (from google-auth>=1.0.0->google-cloud-bigquery==0.30)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python2.7/dist-packages (from requests>=2.18.0->google-cloud-bigquery==0.30)\n",
      "Requirement already satisfied: chardet<3.1.0,>=3.0.2 in /usr/local/lib/python2.7/dist-packages (from requests>=2.18.0->google-cloud-bigquery==0.30)\n",
      "Requirement already satisfied: idna<2.7,>=2.5 in /usr/local/lib/python2.7/dist-packages (from requests>=2.18.0->google-cloud-bigquery==0.30)\n",
      "Requirement already satisfied: urllib3<1.23,>=1.21.1 in /usr/local/lib/python2.7/dist-packages (from requests>=2.18.0->google-cloud-bigquery==0.30)\n",
      "Requirement already satisfied: setuptools>=34.0.0 in /usr/local/lib/python2.7/dist-packages (from google-api-core<0.2.0dev,>=0.1.1->google-cloud-bigquery==0.30)\n",
      "Requirement already satisfied: protobuf>=3.0.0 in /usr/local/lib/python2.7/dist-packages (from google-api-core<0.2.0dev,>=0.1.1->google-cloud-bigquery==0.30)\n",
      "Requirement already satisfied: googleapis-common-protos<2.0dev,>=1.5.3 in /usr/local/lib/python2.7/dist-packages (from google-api-core<0.2.0dev,>=0.1.1->google-cloud-bigquery==0.30)\n",
      "Requirement already satisfied: futures>=3.2.0; python_version < \"3.2\" in /usr/local/lib/python2.7/dist-packages (from google-api-core<0.2.0dev,>=0.1.1->google-cloud-bigquery==0.30)\n",
      "Requirement already satisfied: pytz in /usr/local/lib/python2.7/dist-packages (from google-api-core<0.2.0dev,>=0.1.1->google-cloud-bigquery==0.30)\n",
      "Requirement already satisfied: pyasn1>=0.1.3 in /usr/local/lib/python2.7/dist-packages (from rsa>=3.1.4->google-auth>=1.0.0->google-cloud-bigquery==0.30)\n"
     ]
    }
   ],
   "source": [
    "# My Datalab instance came with an older version and pre 1.0 releases\n",
    "# are not necessarily backwards compatible.\n",
    "!pip install google-cloud-bigquery==0.30"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from google.cloud import bigquery as bq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The next thing you need to know is how to access the specific tables you are interested in.  BigQuery tables are organized into datasets, and datasets are owned by a specific GCP project.  The tables we will be working with in this notebook are in a dataset called **`tcga_201607_beta`**, owned by the **`isb-cgc`** project.  A full table identifier is of the form `<project_id>:<dataset_id>.<table_id>`.  Let's start by getting some basic information about the tables in this dataset:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "client = bq.Client()\n",
    "dataset_ref = client.dataset('tcga_201607_beta', project='isb-cgc')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      6322 rows       1729204 bytes   Annotations\n",
      "     23797 rows       6382147 bytes   Biospecimen_data\n",
      "     11160 rows       4201379 bytes   Clinical_data\n",
      "   2646095 rows     333774244 bytes   Copy_Number_segments\n",
      "3944304319 rows  445303830985 bytes   DNA_Methylation_betas\n",
      " 382335670 rows   43164264006 bytes   DNA_Methylation_chr1\n",
      " 197519895 rows   22301345198 bytes   DNA_Methylation_chr10\n",
      " 235823572 rows   26623975945 bytes   DNA_Methylation_chr11\n",
      " 198050739 rows   22359642619 bytes   DNA_Methylation_chr12\n",
      "  97301675 rows   10986815862 bytes   DNA_Methylation_chr13\n",
      " 123239379 rows   13913712352 bytes   DNA_Methylation_chr14\n",
      " 124566185 rows   14064712239 bytes   DNA_Methylation_chr15\n",
      " 179772812 rows   20296128173 bytes   DNA_Methylation_chr16\n",
      " 234003341 rows   26417830751 bytes   DNA_Methylation_chr17\n",
      "  50216619 rows    5669139362 bytes   DNA_Methylation_chr18\n",
      " 211386795 rows   23862583107 bytes   DNA_Methylation_chr19\n",
      " 279668485 rows   31577200462 bytes   DNA_Methylation_chr2\n",
      "  86858120 rows    9805923353 bytes   DNA_Methylation_chr20\n",
      "  35410447 rows    3997986812 bytes   DNA_Methylation_chr21\n",
      "  70676468 rows    7978947938 bytes   DNA_Methylation_chr22\n",
      " 201119616 rows   22705358910 bytes   DNA_Methylation_chr3\n",
      " 159148744 rows   17968482285 bytes   DNA_Methylation_chr4\n",
      " 195864180 rows   22113162401 bytes   DNA_Methylation_chr5\n",
      " 290275524 rows   32772371379 bytes   DNA_Methylation_chr6\n",
      " 240010275 rows   27097948808 bytes   DNA_Methylation_chr7\n",
      " 164810092 rows   18607886221 bytes   DNA_Methylation_chr8\n",
      "  81260723 rows    9173717922 bytes   DNA_Methylation_chr9\n",
      "  98082681 rows   11072059468 bytes   DNA_Methylation_chrX\n",
      "   2330426 rows     263109775 bytes   DNA_Methylation_chrY\n",
      "   1867233 rows     207365611 bytes   Protein_RPPA_data\n",
      "   5356089 rows    5715538107 bytes   Somatic_Mutation_calls\n",
      "   5738048 rows     657855993 bytes   mRNA_BCGSC_GA_RPKM\n",
      "  38299138 rows    4459086535 bytes   mRNA_BCGSC_HiSeq_RPKM\n",
      "  16794358 rows    1934755686 bytes   mRNA_UNC_GA_RSEM\n",
      " 211284521 rows   24942992190 bytes   mRNA_UNC_HiSeq_RSEM\n",
      " 228078879 rows   26877747876 bytes   mRNA_UNC_RSEM\n",
      "    943447 rows     107533176 bytes   miRNA_BCGSC_GA_EXPR_GroupedByMIMAT\n",
      "   6007080 rows    1003468666 bytes   miRNA_BCGSC_GA_isoform\n",
      "   2614260 rows     306210930 bytes   miRNA_BCGSC_GA_mirna\n",
      "    738920 rows      84104348 bytes   miRNA_BCGSC_GA_mirna3\n",
      "  46543230 rows    7901172305 bytes   miRNA_BCGSC_HiSeq_isoform\n",
      "  16022160 rows    1922031900 bytes   miRNA_BCGSC_HiSeq_mirna\n",
      "   6180663 rows     721224250 bytes   miRNA_BCGSC_HiSeq_mirna3\n",
      "  13226928 rows    1613843246 bytes   miRNA_Expression\n"
     ]
    }
   ],
   "source": [
    "for t in client.list_tables(dataset_ref):\n",
    "  table_ref = dataset_ref.table(t.table_id)\n",
    "  table = client.get_table(table_ref)\n",
    "  print '%10d rows  %12d bytes   %s' \\\n",
    "      % (table.num_rows, table.num_bytes, t.table_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this tutorial, we are going to look at a few different ways that we can use the information in these tables to create cohorts.  Now, you maybe asking what we mean by \"cohort\" and why you might be interested in *creating* one, or maybe what it even means to \"create\" a cohort.  The TCGA dataset includes clinical, biospecimen, and molecular data from over 10,000 cancer patients who agreed to be a part of this landmark research project to build [The Cancer Genome Atlas](http://cancergenome.nih.gov/).  This large dataset was originally organized and studied according to [cancer type](http://cancergenome.nih.gov/cancersselected) but now that this multi-year project is nearing completion, with over 30 types of cancer and over 10,000 tumors analyzed, **you** have the opportunity to look at this dataset from whichever angle most interests you.  Maybe you are particularly interested in early-onset cancers, or gastro-intestinal cancers, or a specific type of genetic mutation.  This is where the idea of a \"cohort\" comes in.  The original TCGA \"cohorts\" were based on cancer type (aka \"study\"), but now you can define a cohort based on virtually any clinical or molecular feature by querying these BigQuery tables.  A cohort is simply a list of samples, using the [TCGA barcode](https://wiki.nci.nih.gov/display/TCGA/TCGA+barcode) system.  Once you have created a cohort you can use it in any number of ways: you could further explore the data available for one cohort, or compare one cohort to another, for example.\n",
    "\n",
    "In the rest of this tutorial, we will create several different cohorts based on different motivating research questions.  We hope that these examples will provide you with a starting point from which you can build, to answer your own research questions."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exploring the Clinical data table\n",
    "Let's start by looking at the clinical data table.  The TCGA dataset contains a few very basic clinical data elements for almost all patients, and contains additional information for some tumor types only.  For example smoking history information is generally available only for lung cancer patients, and BMI (body mass index) is only available for tumor types where that is a known significant risk factor.  Let's take a look at the clinical data table and see how many different pieces of information are available to us:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bqsv\" id=\"11_151908457407\"></div>\n",
       "    <script src=\"/static/components/requirejs/require.js\"></script>\n",
       "    <script>\n",
       "      require.config({\n",
       "        paths: {\n",
       "          base: '/static/base',\n",
       "        },\n",
       "        map: {\n",
       "          '*': {\n",
       "            datalab: 'nbextensions/gcpdatalab'\n",
       "          }\n",
       "        },\n",
       "      });\n",
       "\n",
       "      require(['datalab/bigquery', 'datalab/element!11_151908457407',\n",
       "          'datalab/style!/nbextensions/gcpdatalab/bigquery.css'],\n",
       "        function(bq, dom) {\n",
       "          bq.renderSchema(dom, [{\"type\": \"STRING\", \"name\": \"ParticipantBarcode\"}, {\"type\": \"STRING\", \"name\": \"Study\"}, {\"type\": \"STRING\", \"name\": \"Project\"}, {\"type\": \"STRING\", \"name\": \"ParticipantUUID\"}, {\"type\": \"STRING\", \"name\": \"TSSCode\"}, {\"type\": \"INTEGER\", \"name\": \"age_at_initial_pathologic_diagnosis\"}, {\"type\": \"STRING\", \"name\": \"anatomic_neoplasm_subdivision\"}, {\"type\": \"INTEGER\", \"name\": \"batch_number\"}, {\"type\": \"STRING\", \"name\": \"bcr\"}, {\"type\": \"STRING\", \"name\": \"clinical_M\"}, {\"type\": \"STRING\", \"name\": \"clinical_N\"}, {\"type\": \"STRING\", \"name\": \"clinical_T\"}, {\"type\": \"STRING\", \"name\": \"clinical_stage\"}, {\"type\": \"STRING\", \"name\": \"colorectal_cancer\"}, {\"type\": \"STRING\", \"name\": \"country\"}, {\"type\": \"STRING\", \"name\": \"vital_status\"}, {\"type\": \"INTEGER\", \"name\": \"days_to_birth\"}, {\"type\": \"INTEGER\", \"name\": \"days_to_death\"}, {\"type\": \"INTEGER\", \"name\": \"days_to_last_known_alive\"}, {\"type\": \"INTEGER\", \"name\": \"days_to_last_followup\"}, {\"type\": \"INTEGER\", \"name\": \"days_to_initial_pathologic_diagnosis\"}, {\"type\": \"INTEGER\", \"name\": \"days_to_submitted_specimen_dx\"}, {\"type\": \"STRING\", \"name\": \"ethnicity\"}, {\"type\": \"STRING\", \"name\": \"gender\"}, {\"type\": \"FLOAT\", \"name\": \"gleason_score_combined\"}, {\"type\": \"STRING\", \"name\": \"histological_type\"}, {\"type\": \"STRING\", \"name\": \"history_of_colon_polyps\"}, {\"type\": \"STRING\", \"name\": \"history_of_neoadjuvant_treatment\"}, {\"type\": \"STRING\", \"name\": \"hpv_calls\"}, {\"type\": \"STRING\", \"name\": \"hpv_status\"}, {\"type\": \"STRING\", \"name\": \"icd_10\"}, {\"type\": \"STRING\", \"name\": \"icd_o_3_histology\"}, {\"type\": \"STRING\", \"name\": \"icd_o_3_site\"}, {\"type\": \"STRING\", \"name\": \"lymphatic_invasion\"}, {\"type\": \"STRING\", \"name\": \"lymphnodes_examined\"}, {\"type\": \"STRING\", \"name\": \"lymphovascular_invasion_present\"}, {\"type\": \"STRING\", \"name\": \"menopause_status\"}, {\"type\": \"STRING\", \"name\": \"mononucleotide_and_dinucleotide_marker_panel_analysis_status\"}, {\"type\": \"STRING\", \"name\": \"neoplasm_histologic_grade\"}, {\"type\": \"STRING\", \"name\": \"new_tumor_event_after_initial_treatment\"}, {\"type\": \"FLOAT\", \"name\": \"number_of_lymphnodes_examined\"}, {\"type\": \"FLOAT\", \"name\": \"number_of_lymphnodes_positive_by_he\"}, {\"type\": \"FLOAT\", \"name\": \"number_pack_years_smoked\"}, {\"type\": \"INTEGER\", \"name\": \"year_of_initial_pathologic_diagnosis\"}, {\"type\": \"STRING\", \"name\": \"pathologic_M\"}, {\"type\": \"STRING\", \"name\": \"pathologic_N\"}, {\"type\": \"STRING\", \"name\": \"pathologic_T\"}, {\"type\": \"STRING\", \"name\": \"pathologic_stage\"}, {\"type\": \"STRING\", \"name\": \"person_neoplasm_cancer_status\"}, {\"type\": \"STRING\", \"name\": \"pregnancies\"}, {\"type\": \"STRING\", \"name\": \"primary_neoplasm_melanoma_dx\"}, {\"type\": \"STRING\", \"name\": \"primary_therapy_outcome_success\"}, {\"type\": \"FLOAT\", \"name\": \"psa_value\"}, {\"type\": \"STRING\", \"name\": \"race\"}, {\"type\": \"STRING\", \"name\": \"residual_tumor\"}, {\"type\": \"STRING\", \"name\": \"tobacco_smoking_history\"}, {\"type\": \"STRING\", \"name\": \"tumor_tissue_site\"}, {\"type\": \"STRING\", \"name\": \"tumor_type\"}, {\"type\": \"FLOAT\", \"name\": \"weight\"}, {\"type\": \"FLOAT\", \"name\": \"height\"}, {\"type\": \"FLOAT\", \"name\": \"BMI\"}, {\"type\": \"STRING\", \"name\": \"age_began_smoking_in_years\"}, {\"type\": \"STRING\", \"name\": \"h_pylori_infection\"}, {\"type\": \"STRING\", \"name\": \"other_dx\"}, {\"type\": \"STRING\", \"name\": \"other_malignancy_anatomic_site\"}, {\"type\": \"STRING\", \"name\": \"other_malignancy_histological_type\"}, {\"type\": \"STRING\", \"name\": \"other_malignancy_malignancy_type\"}, {\"type\": \"STRING\", \"name\": \"stopped_smoking_year\"}, {\"type\": \"STRING\", \"name\": \"venous_invasion\"}, {\"type\": \"STRING\", \"name\": \"year_of_tobacco_smoking_onset\"}]);\n",
       "        }\n",
       "      );\n",
       "    </script>\n",
       "    "
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 193,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bq tables describe --name isb-cgc.tcga_201607_beta.Clinical_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "That's a lot of fields!  We can also get at the schema programmatically:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " This table has 70 fields. \n",
      " The first few field names and types are: \n",
      "      [u'ParticipantBarcode', u'Study', u'Project', u'ParticipantUUID', u'TSSCode']\n",
      "      [u'STRING', u'STRING', u'STRING', u'STRING', u'STRING']\n"
     ]
    }
   ],
   "source": [
    "from google.cloud.exceptions import NotFound\n",
    "\n",
    "table_ref = dataset_ref.table('Clinical_data')\n",
    "\n",
    "try:\n",
    "  table = client.get_table(table_ref)\n",
    "  \n",
    "  fieldNames = map(lambda tsf: tsf.name, table.schema)\n",
    "  fieldTypes = map(lambda tsf: tsf.field_type, table.schema)\n",
    "  print \" This table has %d fields. \" % ( len(fieldNames) )\n",
    "  print \" The first few field names and types are: \" \n",
    "  print \"     \", fieldNames[:5]\n",
    "  print \"     \", fieldTypes[:5]\n",
    "\n",
    "except NotFound:\n",
    "  print \" There is no existing table called %s:%s.%s\" % ( table_ref.project, table_ref.dataset_id, table_ref.table_id )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's look at these fields and see which ones might be the most \"interesting\", by looking at how many times they are filled-in (not NULL), or how much variation exists in the values.  If we wanted to look at just a single field, \"tobacco_smoking_history\" for example, we could use a very simple query to get a basic summary:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bqtv\" id=\"12_151908457637\"><table><tr><th>tobacco_smoking_history</th><th>n</th></tr><tr><td>&nbsp;</td><td>8161</td></tr><tr><td>1</td><td>865</td></tr><tr><td>4</td><td>799</td></tr><tr><td>2</td><td>710</td></tr><tr><td>3</td><td>568</td></tr><tr><td>5</td><td>57</td></tr></table></div>\n",
       "    <br />(rows: 6, time: 0.4s, cached, job: job_9RzAuOQcENDMSf6Y6vlEb5lWqzhI)<br />\n",
       "    <script src=\"/static/components/requirejs/require.js\"></script>\n",
       "    <script>\n",
       "      require.config({\n",
       "        paths: {\n",
       "          base: '/static/base',\n",
       "          d3: '//cdnjs.cloudflare.com/ajax/libs/d3/3.4.13/d3',\n",
       "          plotly: 'https://cdn.plot.ly/plotly-1.5.1.min.js?noext',\n",
       "          jquery: '//ajax.googleapis.com/ajax/libs/jquery/2.0.0/jquery.min'\n",
       "        },\n",
       "        map: {\n",
       "          '*': {\n",
       "            datalab: 'nbextensions/gcpdatalab'\n",
       "          }\n",
       "        },\n",
       "        shim: {\n",
       "          plotly: {\n",
       "            deps: ['d3', 'jquery'],\n",
       "            exports: 'plotly'\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "\n",
       "      require(['datalab/charting', 'datalab/element!12_151908457637', 'base/js/events',\n",
       "          'datalab/style!/nbextensions/gcpdatalab/charting.css'],\n",
       "        function(charts, dom, events) {\n",
       "          charts.render('gcharts', dom, events, 'table', [], {\"rows\": [{\"c\": [{\"v\": null}, {\"v\": 8161}]}, {\"c\": [{\"v\": \"1\"}, {\"v\": 865}]}, {\"c\": [{\"v\": \"4\"}, {\"v\": 799}]}, {\"c\": [{\"v\": \"2\"}, {\"v\": 710}]}, {\"c\": [{\"v\": \"3\"}, {\"v\": 568}]}, {\"c\": [{\"v\": \"5\"}, {\"v\": 57}]}], \"cols\": [{\"type\": \"string\", \"id\": \"tobacco_smoking_history\", \"label\": \"tobacco_smoking_history\"}, {\"type\": \"number\", \"id\": \"n\", \"label\": \"n\"}]},\n",
       "            {\n",
       "              pageSize: 25,\n",
       "              cssClassNames:  {\n",
       "                tableRow: 'gchart-table-row',\n",
       "                headerRow: 'gchart-table-headerrow',\n",
       "                oddTableRow: 'gchart-table-oddrow',\n",
       "                selectedTableRow: 'gchart-table-selectedrow',\n",
       "                hoverTableRow: 'gchart-table-hoverrow',\n",
       "                tableCell: 'gchart-table-cell',\n",
       "                headerCell: 'gchart-table-headercell',\n",
       "                rowNumberCell: 'gchart-table-rownumcell'\n",
       "              }\n",
       "            },\n",
       "            {source_index: 1, fields: 'tobacco_smoking_history,n'},\n",
       "            0,\n",
       "            6);\n",
       "        }\n",
       "      );\n",
       "    </script>\n",
       "  "
      ],
      "text/plain": [
       "QueryResultsTable job_9RzAuOQcENDMSf6Y6vlEb5lWqzhI"
      ]
     },
     "execution_count": 195,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bq query\n",
    "SELECT tobacco_smoking_history, COUNT(*) AS n\n",
    "FROM `isb-cgc.tcga_201607_beta.Clinical_data`\n",
    "GROUP BY tobacco_smoking_history\n",
    "ORDER BY n DESC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "But if we want to loop over *all* fields and get a sense of which fields might provide us with useful criteria for specifying a cohort, we'll want to automate that.  We'll put a threshold on the minimum number of patients that we expect information for, and the maximum number of unique values (since fields such as the \"ParticipantBarcode\" will be unique for every patient and, although we will need that field later, it's probably not useful for defining a cohort)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " The Clinical_data table describes a total of 11160 patients. \n",
      "     > Study has 11160 values with 33 unique (BRCA occurs 1097 times) \n",
      "     > age_at_initial_pathologic_diagnosis has 11109 values (mean=59, sigma=14) \n",
      "     > batch_number has 11160 values (mean=203, sigma=135) \n",
      "     > vital_status has 11156 values with 2 unique (Alive occurs 7534 times) \n",
      "     > days_to_birth has 11041 values (mean=-21763, sigma=5266) \n",
      "     > days_to_last_known_alive has 11102 values (mean=1037, sigma=1041) \n",
      "     > gender has 11160 values with 2 unique (FEMALE occurs 5815 times) \n",
      "     > year_of_initial_pathologic_diagnosis has 11030 values (mean=2008, sigma=4) \n",
      "     > person_neoplasm_cancer_status has 10236 values with 2 unique (TUMOR FREE occurs 6507 times) \n",
      "     > race has 9835 values with 5 unique (WHITE occurs 8186 times) \n",
      " \n",
      " Found 10 potentially interesting features: \n",
      "    [u'Study', u'age_at_initial_pathologic_diagnosis', u'batch_number', u'vital_status', u'days_to_birth', u'days_to_last_known_alive', u'gender', u'year_of_initial_pathologic_diagnosis', u'person_neoplasm_cancer_status', u'race']\n"
     ]
    }
   ],
   "source": [
    "numPatients = table.num_rows\n",
    "print \" The %s table describes a total of %d patients. \" % ( table.table_id, numPatients )\n",
    "\n",
    "# let's set a threshold for the minimum number of values that a field should have,\n",
    "# and also the maximum number of unique values\n",
    "minNumPatients = int(numPatients*0.80)\n",
    "maxNumValues = 50\n",
    "\n",
    "numInteresting = 0\n",
    "iList = []\n",
    "for iField in range(len(fieldNames)):\n",
    "  aField = fieldNames[iField]\n",
    "  aType = fieldTypes[iField]\n",
    "  try:\n",
    "    qString = \"SELECT {0} FROM `isb-cgc.tcga_201607_beta.Clinical_data`\".format(aField)\n",
    "    df = client.query(qString).result().to_dataframe()\n",
    "    summary = df[str(aField)].describe()\n",
    "    if ( aType == \"STRING\" ):\n",
    "      topFrac = float(summary['freq'])/float(summary['count'])\n",
    "      if ( summary['count'] >= minNumPatients ):\n",
    "        if ( summary['unique'] <= maxNumValues and summary['unique'] > 1 ):\n",
    "          if ( topFrac < 0.90 ):\n",
    "            numInteresting += 1\n",
    "            iList += [aField]\n",
    "            print \"     > %s has %d values with %d unique (%s occurs %d times) \" \\\n",
    "              % (str(aField), summary['count'], summary['unique'], summary['top'], summary['freq'])\n",
    "    else:\n",
    "      if ( summary['count'] >= minNumPatients ):\n",
    "        if ( summary['std'] > 0.1 ):\n",
    "          numInteresting += 1\n",
    "          iList += [aField]\n",
    "          print \"     > %s has %d values (mean=%.0f, sigma=%.0f) \" \\\n",
    "            % (str(aField), summary['count'], summary['mean'], summary['std'])\n",
    "  except:\n",
    "    pass\n",
    "\n",
    "print \" \"\n",
    "print \" Found %d potentially interesting features: \" % numInteresting\n",
    "print \"   \", iList"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The above helps us narrow down on which fields are likely to be the most useful, but if you have a specific interest, for example in menopause or HPV status, you can still look at those in more detail very easily: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bqtv\" id=\"13_151908492201\"><table><tr><th>menopause_status</th><th>n</th></tr><tr><td>Post (prior bilateral ovariectomy OR &gt;12 mo since LMP with no prior hysterectomy)</td><td>1291</td></tr><tr><td>Pre (&lt;6 months since LMP AND no prior bilateral ovariectomy AND not on estrogen replacement)</td><td>389</td></tr><tr><td>Peri (6-12 months since last menstrual period)</td><td>82</td></tr><tr><td>Indeterminate (neither Pre or Postmenopausal)</td><td>54</td></tr></table></div>\n",
       "    <br />(rows: 4, time: 1.0s, cached, job: job_J5zPkY_hUWvb1rlx8ioCdDaeTNU1)<br />\n",
       "    <script src=\"/static/components/requirejs/require.js\"></script>\n",
       "    <script>\n",
       "      require.config({\n",
       "        paths: {\n",
       "          base: '/static/base',\n",
       "          d3: '//cdnjs.cloudflare.com/ajax/libs/d3/3.4.13/d3',\n",
       "          plotly: 'https://cdn.plot.ly/plotly-1.5.1.min.js?noext',\n",
       "          jquery: '//ajax.googleapis.com/ajax/libs/jquery/2.0.0/jquery.min'\n",
       "        },\n",
       "        map: {\n",
       "          '*': {\n",
       "            datalab: 'nbextensions/gcpdatalab'\n",
       "          }\n",
       "        },\n",
       "        shim: {\n",
       "          plotly: {\n",
       "            deps: ['d3', 'jquery'],\n",
       "            exports: 'plotly'\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "\n",
       "      require(['datalab/charting', 'datalab/element!13_151908492201', 'base/js/events',\n",
       "          'datalab/style!/nbextensions/gcpdatalab/charting.css'],\n",
       "        function(charts, dom, events) {\n",
       "          charts.render('gcharts', dom, events, 'table', [], {\"rows\": [{\"c\": [{\"v\": \"Post (prior bilateral ovariectomy OR >12 mo since LMP with no prior hysterectomy)\"}, {\"v\": 1291}]}, {\"c\": [{\"v\": \"Pre (<6 months since LMP AND no prior bilateral ovariectomy AND not on estrogen replacement)\"}, {\"v\": 389}]}, {\"c\": [{\"v\": \"Peri (6-12 months since last menstrual period)\"}, {\"v\": 82}]}, {\"c\": [{\"v\": \"Indeterminate (neither Pre or Postmenopausal)\"}, {\"v\": 54}]}], \"cols\": [{\"type\": \"string\", \"id\": \"menopause_status\", \"label\": \"menopause_status\"}, {\"type\": \"number\", \"id\": \"n\", \"label\": \"n\"}]},\n",
       "            {\n",
       "              pageSize: 25,\n",
       "              cssClassNames:  {\n",
       "                tableRow: 'gchart-table-row',\n",
       "                headerRow: 'gchart-table-headerrow',\n",
       "                oddTableRow: 'gchart-table-oddrow',\n",
       "                selectedTableRow: 'gchart-table-selectedrow',\n",
       "                hoverTableRow: 'gchart-table-hoverrow',\n",
       "                tableCell: 'gchart-table-cell',\n",
       "                headerCell: 'gchart-table-headercell',\n",
       "                rowNumberCell: 'gchart-table-rownumcell'\n",
       "              }\n",
       "            },\n",
       "            {source_index: 8, fields: 'menopause_status,n'},\n",
       "            0,\n",
       "            4);\n",
       "        }\n",
       "      );\n",
       "    </script>\n",
       "  "
      ],
      "text/plain": [
       "QueryResultsTable job_J5zPkY_hUWvb1rlx8ioCdDaeTNU1"
      ]
     },
     "execution_count": 198,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bq query\n",
    "SELECT menopause_status, COUNT(*) AS n\n",
    "FROM `isb-cgc.tcga_201607_beta.Clinical_data`\n",
    "WHERE menopause_status IS NOT NULL\n",
    "GROUP BY menopause_status\n",
    "ORDER BY n DESC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We might wonder which specific tumor types have menopause information:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bqtv\" id=\"14_151908492442\"><table><tr><th>Study</th><th>n</th></tr><tr><td>BRCA</td><td>1007</td></tr><tr><td>UCEC</td><td>517</td></tr><tr><td>CESC</td><td>237</td></tr><tr><td>UCS</td><td>55</td></tr></table></div>\n",
       "    <br />(rows: 4, time: 0.2s, cached, job: job_aO5izPHcenh1sxqMtdgy6VzOtSjE)<br />\n",
       "    <script src=\"/static/components/requirejs/require.js\"></script>\n",
       "    <script>\n",
       "      require.config({\n",
       "        paths: {\n",
       "          base: '/static/base',\n",
       "          d3: '//cdnjs.cloudflare.com/ajax/libs/d3/3.4.13/d3',\n",
       "          plotly: 'https://cdn.plot.ly/plotly-1.5.1.min.js?noext',\n",
       "          jquery: '//ajax.googleapis.com/ajax/libs/jquery/2.0.0/jquery.min'\n",
       "        },\n",
       "        map: {\n",
       "          '*': {\n",
       "            datalab: 'nbextensions/gcpdatalab'\n",
       "          }\n",
       "        },\n",
       "        shim: {\n",
       "          plotly: {\n",
       "            deps: ['d3', 'jquery'],\n",
       "            exports: 'plotly'\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "\n",
       "      require(['datalab/charting', 'datalab/element!14_151908492442', 'base/js/events',\n",
       "          'datalab/style!/nbextensions/gcpdatalab/charting.css'],\n",
       "        function(charts, dom, events) {\n",
       "          charts.render('gcharts', dom, events, 'table', [], {\"rows\": [{\"c\": [{\"v\": \"BRCA\"}, {\"v\": 1007}]}, {\"c\": [{\"v\": \"UCEC\"}, {\"v\": 517}]}, {\"c\": [{\"v\": \"CESC\"}, {\"v\": 237}]}, {\"c\": [{\"v\": \"UCS\"}, {\"v\": 55}]}], \"cols\": [{\"type\": \"string\", \"id\": \"Study\", \"label\": \"Study\"}, {\"type\": \"number\", \"id\": \"n\", \"label\": \"n\"}]},\n",
       "            {\n",
       "              pageSize: 25,\n",
       "              cssClassNames:  {\n",
       "                tableRow: 'gchart-table-row',\n",
       "                headerRow: 'gchart-table-headerrow',\n",
       "                oddTableRow: 'gchart-table-oddrow',\n",
       "                selectedTableRow: 'gchart-table-selectedrow',\n",
       "                hoverTableRow: 'gchart-table-hoverrow',\n",
       "                tableCell: 'gchart-table-cell',\n",
       "                headerCell: 'gchart-table-headercell',\n",
       "                rowNumberCell: 'gchart-table-rownumcell'\n",
       "              }\n",
       "            },\n",
       "            {source_index: 9, fields: 'Study,n'},\n",
       "            0,\n",
       "            4);\n",
       "        }\n",
       "      );\n",
       "    </script>\n",
       "  "
      ],
      "text/plain": [
       "QueryResultsTable job_aO5izPHcenh1sxqMtdgy6VzOtSjE"
      ]
     },
     "execution_count": 199,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bq query\n",
    "SELECT Study, COUNT(*) AS n\n",
    "FROM `isb-cgc.tcga_201607_beta.Clinical_data`\n",
    "WHERE menopause_status IS NOT NULL\n",
    "GROUP BY Study\n",
    "ORDER BY n DESC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bqtv\" id=\"15_151908492657\"><table><tr><th>hpv_status</th><th>hpv_calls</th><th>n</th></tr><tr><td>Negative</td><td>&nbsp;</td><td>664</td></tr><tr><td>Positive</td><td>HPV16</td><td>238</td></tr><tr><td>Positive</td><td>HPV18</td><td>41</td></tr><tr><td>Positive</td><td>HPV33</td><td>25</td></tr><tr><td>Positive</td><td>HPV45</td><td>24</td></tr></table></div>\n",
       "    <br />(rows: 5, time: 0.3s, cached, job: job_qUamj-dRiBi8AQr-UxR326DXmIBx)<br />\n",
       "    <script src=\"/static/components/requirejs/require.js\"></script>\n",
       "    <script>\n",
       "      require.config({\n",
       "        paths: {\n",
       "          base: '/static/base',\n",
       "          d3: '//cdnjs.cloudflare.com/ajax/libs/d3/3.4.13/d3',\n",
       "          plotly: 'https://cdn.plot.ly/plotly-1.5.1.min.js?noext',\n",
       "          jquery: '//ajax.googleapis.com/ajax/libs/jquery/2.0.0/jquery.min'\n",
       "        },\n",
       "        map: {\n",
       "          '*': {\n",
       "            datalab: 'nbextensions/gcpdatalab'\n",
       "          }\n",
       "        },\n",
       "        shim: {\n",
       "          plotly: {\n",
       "            deps: ['d3', 'jquery'],\n",
       "            exports: 'plotly'\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "\n",
       "      require(['datalab/charting', 'datalab/element!15_151908492657', 'base/js/events',\n",
       "          'datalab/style!/nbextensions/gcpdatalab/charting.css'],\n",
       "        function(charts, dom, events) {\n",
       "          charts.render('gcharts', dom, events, 'table', [], {\"rows\": [{\"c\": [{\"v\": \"Negative\"}, {\"v\": null}, {\"v\": 664}]}, {\"c\": [{\"v\": \"Positive\"}, {\"v\": \"HPV16\"}, {\"v\": 238}]}, {\"c\": [{\"v\": \"Positive\"}, {\"v\": \"HPV18\"}, {\"v\": 41}]}, {\"c\": [{\"v\": \"Positive\"}, {\"v\": \"HPV33\"}, {\"v\": 25}]}, {\"c\": [{\"v\": \"Positive\"}, {\"v\": \"HPV45\"}, {\"v\": 24}]}], \"cols\": [{\"type\": \"string\", \"id\": \"hpv_status\", \"label\": \"hpv_status\"}, {\"type\": \"string\", \"id\": \"hpv_calls\", \"label\": \"hpv_calls\"}, {\"type\": \"number\", \"id\": \"n\", \"label\": \"n\"}]},\n",
       "            {\n",
       "              pageSize: 25,\n",
       "              cssClassNames:  {\n",
       "                tableRow: 'gchart-table-row',\n",
       "                headerRow: 'gchart-table-headerrow',\n",
       "                oddTableRow: 'gchart-table-oddrow',\n",
       "                selectedTableRow: 'gchart-table-selectedrow',\n",
       "                hoverTableRow: 'gchart-table-hoverrow',\n",
       "                tableCell: 'gchart-table-cell',\n",
       "                headerCell: 'gchart-table-headercell',\n",
       "                rowNumberCell: 'gchart-table-rownumcell'\n",
       "              }\n",
       "            },\n",
       "            {source_index: 0, fields: 'hpv_status,hpv_calls,n'},\n",
       "            0,\n",
       "            5);\n",
       "        }\n",
       "      );\n",
       "    </script>\n",
       "  "
      ],
      "text/plain": [
       "QueryResultsTable job_qUamj-dRiBi8AQr-UxR326DXmIBx"
      ]
     },
     "execution_count": 200,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bq query\n",
    "SELECT hpv_status, hpv_calls, COUNT(*) AS n\n",
    "FROM `isb-cgc.tcga_201607_beta.Clinical_data`\n",
    "WHERE hpv_status IS NOT NULL\n",
    "GROUP BY hpv_status, hpv_calls\n",
    "HAVING n > 20\n",
    "ORDER BY n DESC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TCGA Annotations\n",
    "\n",
    "An additional factor to consider, when creating a cohort is that there may be additional information that might lead one to exclude a particular patient from a cohort.  In certain instances, patients have been redacted or excluded from analyses for reasons such as prior treatment, etc, but since different researchers may have different criteria for using or excluding certain patients or certain samples from their analyses, in many cases the data is still available while at the same time \"annotations\" may have been entered into a searchable [database](https://tcga-data.nci.nih.gov/annotations/).  These annotations have also been uploaded into a BigQuery table and can be used in conjuction with the other BigQuery tables."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Early-onset Breast Cancer\n",
    "\n",
    "Now that we have a better idea of what types of information is available in the Clinical data table, let's create a cohort consisting of female breast-cancer patients, diagnosed at the age of 50 or younger."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this next code cell, we define several queries within a **`module`** which allows us to use them both individually and by reference in the final, main query.  \n",
    "+ the first query, called **`select_on_annotations`**, finds all patients in the Annotations table which have either been 'redacted' or had 'unacceptable prior treatment';  \n",
    "+ the second query, **`select_on_clinical`** selects all female breast-cancer patients who were diagnosed at age 50 or younger, while also pulling out a few additional fields that might be of interest;  and\n",
    "+ the final query joins these two together and returns just those patients that meet the clinical-criteria and do **not** meet the exclusion-criteria."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bq query --name select_on_annotations\n",
    "\n",
    "SELECT\n",
    "  ParticipantBarcode,\n",
    "  annotationCategoryName AS categoryName,\n",
    "  annotationClassification AS classificationName\n",
    "FROM\n",
    "  `isb-cgc.tcga_201607_beta.Annotations`\n",
    "WHERE\n",
    "  ( itemTypeName=\"Patient\"\n",
    "    AND (annotationCategoryName=\"History of unacceptable prior treatment related to a prior/other malignancy\"\n",
    "      OR annotationClassification=\"Redaction\" ) )\n",
    "GROUP BY\n",
    "  ParticipantBarcode,\n",
    "  categoryName,\n",
    "  classificationName"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bq query --name select_on_clinical\n",
    "\n",
    "SELECT\n",
    "  ParticipantBarcode,\n",
    "  vital_status,\n",
    "  days_to_last_known_alive,\n",
    "  ethnicity,\n",
    "  histological_type,\n",
    "  menopause_status,\n",
    "  race\n",
    "FROM\n",
    "  `isb-cgc.tcga_201607_beta.Clinical_data`\n",
    "WHERE\n",
    "  ( Study=\"BRCA\"\n",
    "    AND age_at_initial_pathologic_diagnosis<=50\n",
    "    AND gender=\"FEMALE\" )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bq query --name create_cohort --subqueries select_on_annotations select_on_clinical\n",
    "\n",
    "SELECT\n",
    "  ParticipantBarcode\n",
    "FROM (\n",
    "  SELECT\n",
    "    a.categoryName,\n",
    "    a.classificationName,\n",
    "    c.ParticipantBarcode\n",
    "  FROM select_on_annotations AS a\n",
    "  FULL JOIN select_on_clinical AS c\n",
    "  ON\n",
    "    a.ParticipantBarcode = c.ParticipantBarcode\n",
    "  WHERE\n",
    "    a.ParticipantBarcode IS NOT NULL\n",
    "    OR c.ParticipantBarcode IS NOT NULL\n",
    "  ORDER BY\n",
    "    a.classificationName,\n",
    "    a.categoryName,\n",
    "    c.ParticipantBarcode\n",
    ")\n",
    "WHERE\n",
    "  categoryName IS NULL\n",
    "  AND classificationName IS NULL\n",
    "  AND ParticipantBarcode IS NOT NULL\n",
    "ORDER BY\n",
    "  ParticipantBarcode"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we explicitly call just the first query in the module, and we get a list of 212 patients with one of these disqualifying annotations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ParticipantBarcode</th>\n",
       "      <th>categoryName</th>\n",
       "      <th>classificationName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-24-0981</td>\n",
       "      <td>Duplicate case</td>\n",
       "      <td>Redaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-BR-4205</td>\n",
       "      <td>Duplicate case</td>\n",
       "      <td>Redaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-BR-4282</td>\n",
       "      <td>Duplicate case</td>\n",
       "      <td>Redaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-BR-4273</td>\n",
       "      <td>Duplicate case</td>\n",
       "      <td>Redaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-BR-4285</td>\n",
       "      <td>Duplicate case</td>\n",
       "      <td>Redaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TCGA-BR-4266</td>\n",
       "      <td>Duplicate case</td>\n",
       "      <td>Redaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TCGA-BR-4197</td>\n",
       "      <td>Duplicate case</td>\n",
       "      <td>Redaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>TCGA-BR-4264</td>\n",
       "      <td>Duplicate case</td>\n",
       "      <td>Redaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>TCGA-BR-4276</td>\n",
       "      <td>Duplicate case</td>\n",
       "      <td>Redaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>TCGA-BR-4260</td>\n",
       "      <td>Duplicate case</td>\n",
       "      <td>Redaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>TCGA-BR-4272</td>\n",
       "      <td>Duplicate case</td>\n",
       "      <td>Redaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>TCGA-BR-4196</td>\n",
       "      <td>Duplicate case</td>\n",
       "      <td>Redaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>TCGA-BR-4270</td>\n",
       "      <td>Duplicate case</td>\n",
       "      <td>Redaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>TCGA-BR-4271</td>\n",
       "      <td>Duplicate case</td>\n",
       "      <td>Redaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>TCGA-BR-4286</td>\n",
       "      <td>Duplicate case</td>\n",
       "      <td>Redaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>TCGA-BR-4263</td>\n",
       "      <td>Duplicate case</td>\n",
       "      <td>Redaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>TCGA-BR-4274</td>\n",
       "      <td>Duplicate case</td>\n",
       "      <td>Redaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>TCGA-O2-A5IC</td>\n",
       "      <td>Duplicate case</td>\n",
       "      <td>Redaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>TCGA-06-0748</td>\n",
       "      <td>Genotype mismatch</td>\n",
       "      <td>Redaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>TCGA-PN-A8M9</td>\n",
       "      <td>Genotype mismatch</td>\n",
       "      <td>Redaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>TCGA-66-2752</td>\n",
       "      <td>Genotype mismatch</td>\n",
       "      <td>Redaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>TCGA-02-0117</td>\n",
       "      <td>Genotype mismatch</td>\n",
       "      <td>Redaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>TCGA-02-2488</td>\n",
       "      <td>Genotype mismatch</td>\n",
       "      <td>Redaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>TCGA-66-2746</td>\n",
       "      <td>Genotype mismatch</td>\n",
       "      <td>Redaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>TCGA-66-2751</td>\n",
       "      <td>Genotype mismatch</td>\n",
       "      <td>Redaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>TCGA-08-0384</td>\n",
       "      <td>Genotype mismatch</td>\n",
       "      <td>Redaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>TCGA-66-2747</td>\n",
       "      <td>Genotype mismatch</td>\n",
       "      <td>Redaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>TCGA-E2-A1IP</td>\n",
       "      <td>Genotype mismatch</td>\n",
       "      <td>Redaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>TCGA-14-1036</td>\n",
       "      <td>Genotype mismatch</td>\n",
       "      <td>Redaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>TCGA-66-2750</td>\n",
       "      <td>Genotype mismatch</td>\n",
       "      <td>Redaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>TCGA-CX-7082</td>\n",
       "      <td>History of unacceptable prior treatment relate...</td>\n",
       "      <td>Notification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>TCGA-CG-4460</td>\n",
       "      <td>History of unacceptable prior treatment relate...</td>\n",
       "      <td>Notification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>TCGA-21-1070</td>\n",
       "      <td>History of unacceptable prior treatment relate...</td>\n",
       "      <td>Notification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>TCGA-BD-A3ER</td>\n",
       "      <td>History of unacceptable prior treatment relate...</td>\n",
       "      <td>Notification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>186</th>\n",
       "      <td>TCGA-G4-6625</td>\n",
       "      <td>History of unacceptable prior treatment relate...</td>\n",
       "      <td>Notification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>TCGA-CZ-4865</td>\n",
       "      <td>History of unacceptable prior treatment relate...</td>\n",
       "      <td>Notification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>188</th>\n",
       "      <td>TCGA-38-4630</td>\n",
       "      <td>History of unacceptable prior treatment relate...</td>\n",
       "      <td>Notification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>TCGA-AA-3554</td>\n",
       "      <td>History of unacceptable prior treatment relate...</td>\n",
       "      <td>Notification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>TCGA-EJ-7312</td>\n",
       "      <td>History of unacceptable prior treatment relate...</td>\n",
       "      <td>Notification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>TCGA-BQ-7049</td>\n",
       "      <td>History of unacceptable prior treatment relate...</td>\n",
       "      <td>Notification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>TCGA-G4-6314</td>\n",
       "      <td>History of unacceptable prior treatment relate...</td>\n",
       "      <td>Notification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>TCGA-AA-3516</td>\n",
       "      <td>History of unacceptable prior treatment relate...</td>\n",
       "      <td>Notification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>TCGA-G5-6235</td>\n",
       "      <td>History of unacceptable prior treatment relate...</td>\n",
       "      <td>Notification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>TCGA-AG-3611</td>\n",
       "      <td>History of unacceptable prior treatment relate...</td>\n",
       "      <td>Notification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>TCGA-PC-A5DP</td>\n",
       "      <td>History of unacceptable prior treatment relate...</td>\n",
       "      <td>Notification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>TCGA-AR-A2LR</td>\n",
       "      <td>History of unacceptable prior treatment relate...</td>\n",
       "      <td>Notification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>TCGA-CQ-A4C7</td>\n",
       "      <td>History of unacceptable prior treatment relate...</td>\n",
       "      <td>Notification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>TCGA-BA-4077</td>\n",
       "      <td>History of unacceptable prior treatment relate...</td>\n",
       "      <td>Notification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>TCGA-60-2712</td>\n",
       "      <td>History of unacceptable prior treatment relate...</td>\n",
       "      <td>Notification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>TCGA-AL-A5DJ</td>\n",
       "      <td>History of unacceptable prior treatment relate...</td>\n",
       "      <td>Notification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>TCGA-BP-4976</td>\n",
       "      <td>History of unacceptable prior treatment relate...</td>\n",
       "      <td>Notification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>TCGA-AX-A1CR</td>\n",
       "      <td>History of unacceptable prior treatment relate...</td>\n",
       "      <td>Notification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>TCGA-BG-A0MS</td>\n",
       "      <td>History of unacceptable prior treatment relate...</td>\n",
       "      <td>Notification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>TCGA-CG-4629</td>\n",
       "      <td>History of unacceptable prior treatment relate...</td>\n",
       "      <td>Notification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>TCGA-BP-4988</td>\n",
       "      <td>History of unacceptable prior treatment relate...</td>\n",
       "      <td>Notification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>TCGA-38-4629</td>\n",
       "      <td>History of unacceptable prior treatment relate...</td>\n",
       "      <td>Notification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>TCGA-EY-A1G7</td>\n",
       "      <td>History of unacceptable prior treatment relate...</td>\n",
       "      <td>Notification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>TCGA-BG-A220</td>\n",
       "      <td>History of unacceptable prior treatment relate...</td>\n",
       "      <td>Notification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210</th>\n",
       "      <td>TCGA-18-3409</td>\n",
       "      <td>History of unacceptable prior treatment relate...</td>\n",
       "      <td>Notification</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>TCGA-AG-A00H</td>\n",
       "      <td>History of unacceptable prior treatment relate...</td>\n",
       "      <td>Notification</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>212 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    ParticipantBarcode                                       categoryName  \\\n",
       "0         TCGA-24-0981                                     Duplicate case   \n",
       "1         TCGA-BR-4205                                     Duplicate case   \n",
       "2         TCGA-BR-4282                                     Duplicate case   \n",
       "3         TCGA-BR-4273                                     Duplicate case   \n",
       "4         TCGA-BR-4285                                     Duplicate case   \n",
       "5         TCGA-BR-4266                                     Duplicate case   \n",
       "6         TCGA-BR-4197                                     Duplicate case   \n",
       "7         TCGA-BR-4264                                     Duplicate case   \n",
       "8         TCGA-BR-4276                                     Duplicate case   \n",
       "9         TCGA-BR-4260                                     Duplicate case   \n",
       "10        TCGA-BR-4272                                     Duplicate case   \n",
       "11        TCGA-BR-4196                                     Duplicate case   \n",
       "12        TCGA-BR-4270                                     Duplicate case   \n",
       "13        TCGA-BR-4271                                     Duplicate case   \n",
       "14        TCGA-BR-4286                                     Duplicate case   \n",
       "15        TCGA-BR-4263                                     Duplicate case   \n",
       "16        TCGA-BR-4274                                     Duplicate case   \n",
       "17        TCGA-O2-A5IC                                     Duplicate case   \n",
       "18        TCGA-06-0748                                  Genotype mismatch   \n",
       "19        TCGA-PN-A8M9                                  Genotype mismatch   \n",
       "20        TCGA-66-2752                                  Genotype mismatch   \n",
       "21        TCGA-02-0117                                  Genotype mismatch   \n",
       "22        TCGA-02-2488                                  Genotype mismatch   \n",
       "23        TCGA-66-2746                                  Genotype mismatch   \n",
       "24        TCGA-66-2751                                  Genotype mismatch   \n",
       "25        TCGA-08-0384                                  Genotype mismatch   \n",
       "26        TCGA-66-2747                                  Genotype mismatch   \n",
       "27        TCGA-E2-A1IP                                  Genotype mismatch   \n",
       "28        TCGA-14-1036                                  Genotype mismatch   \n",
       "29        TCGA-66-2750                                  Genotype mismatch   \n",
       "..                 ...                                                ...   \n",
       "182       TCGA-CX-7082  History of unacceptable prior treatment relate...   \n",
       "183       TCGA-CG-4460  History of unacceptable prior treatment relate...   \n",
       "184       TCGA-21-1070  History of unacceptable prior treatment relate...   \n",
       "185       TCGA-BD-A3ER  History of unacceptable prior treatment relate...   \n",
       "186       TCGA-G4-6625  History of unacceptable prior treatment relate...   \n",
       "187       TCGA-CZ-4865  History of unacceptable prior treatment relate...   \n",
       "188       TCGA-38-4630  History of unacceptable prior treatment relate...   \n",
       "189       TCGA-AA-3554  History of unacceptable prior treatment relate...   \n",
       "190       TCGA-EJ-7312  History of unacceptable prior treatment relate...   \n",
       "191       TCGA-BQ-7049  History of unacceptable prior treatment relate...   \n",
       "192       TCGA-G4-6314  History of unacceptable prior treatment relate...   \n",
       "193       TCGA-AA-3516  History of unacceptable prior treatment relate...   \n",
       "194       TCGA-G5-6235  History of unacceptable prior treatment relate...   \n",
       "195       TCGA-AG-3611  History of unacceptable prior treatment relate...   \n",
       "196       TCGA-PC-A5DP  History of unacceptable prior treatment relate...   \n",
       "197       TCGA-AR-A2LR  History of unacceptable prior treatment relate...   \n",
       "198       TCGA-CQ-A4C7  History of unacceptable prior treatment relate...   \n",
       "199       TCGA-BA-4077  History of unacceptable prior treatment relate...   \n",
       "200       TCGA-60-2712  History of unacceptable prior treatment relate...   \n",
       "201       TCGA-AL-A5DJ  History of unacceptable prior treatment relate...   \n",
       "202       TCGA-BP-4976  History of unacceptable prior treatment relate...   \n",
       "203       TCGA-AX-A1CR  History of unacceptable prior treatment relate...   \n",
       "204       TCGA-BG-A0MS  History of unacceptable prior treatment relate...   \n",
       "205       TCGA-CG-4629  History of unacceptable prior treatment relate...   \n",
       "206       TCGA-BP-4988  History of unacceptable prior treatment relate...   \n",
       "207       TCGA-38-4629  History of unacceptable prior treatment relate...   \n",
       "208       TCGA-EY-A1G7  History of unacceptable prior treatment relate...   \n",
       "209       TCGA-BG-A220  History of unacceptable prior treatment relate...   \n",
       "210       TCGA-18-3409  History of unacceptable prior treatment relate...   \n",
       "211       TCGA-AG-A00H  History of unacceptable prior treatment relate...   \n",
       "\n",
       "    classificationName  \n",
       "0            Redaction  \n",
       "1            Redaction  \n",
       "2            Redaction  \n",
       "3            Redaction  \n",
       "4            Redaction  \n",
       "5            Redaction  \n",
       "6            Redaction  \n",
       "7            Redaction  \n",
       "8            Redaction  \n",
       "9            Redaction  \n",
       "10           Redaction  \n",
       "11           Redaction  \n",
       "12           Redaction  \n",
       "13           Redaction  \n",
       "14           Redaction  \n",
       "15           Redaction  \n",
       "16           Redaction  \n",
       "17           Redaction  \n",
       "18           Redaction  \n",
       "19           Redaction  \n",
       "20           Redaction  \n",
       "21           Redaction  \n",
       "22           Redaction  \n",
       "23           Redaction  \n",
       "24           Redaction  \n",
       "25           Redaction  \n",
       "26           Redaction  \n",
       "27           Redaction  \n",
       "28           Redaction  \n",
       "29           Redaction  \n",
       "..                 ...  \n",
       "182       Notification  \n",
       "183       Notification  \n",
       "184       Notification  \n",
       "185       Notification  \n",
       "186       Notification  \n",
       "187       Notification  \n",
       "188       Notification  \n",
       "189       Notification  \n",
       "190       Notification  \n",
       "191       Notification  \n",
       "192       Notification  \n",
       "193       Notification  \n",
       "194       Notification  \n",
       "195       Notification  \n",
       "196       Notification  \n",
       "197       Notification  \n",
       "198       Notification  \n",
       "199       Notification  \n",
       "200       Notification  \n",
       "201       Notification  \n",
       "202       Notification  \n",
       "203       Notification  \n",
       "204       Notification  \n",
       "205       Notification  \n",
       "206       Notification  \n",
       "207       Notification  \n",
       "208       Notification  \n",
       "209       Notification  \n",
       "210       Notification  \n",
       "211       Notification  \n",
       "\n",
       "[212 rows x 3 columns]"
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "select_on_annotations.execute().result().to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "and here we explicitly call just the second query, resulting in 329 patients:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ParticipantBarcode</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>days_to_last_known_alive</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>histological_type</th>\n",
       "      <th>menopause_status</th>\n",
       "      <th>race</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-B6-A0IN</td>\n",
       "      <td>Dead</td>\n",
       "      <td>2573</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Indeterminate (neither Pre or Postmenopausal)</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-B6-A0RU</td>\n",
       "      <td>Alive</td>\n",
       "      <td>8605</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Indeterminate (neither Pre or Postmenopausal)</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-EW-A1IX</td>\n",
       "      <td>Alive</td>\n",
       "      <td>1208</td>\n",
       "      <td>HISPANIC OR LATINO</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Peri (6-12 months since last menstrual period)</td>\n",
       "      <td>ASIAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-AO-A0JG</td>\n",
       "      <td>Alive</td>\n",
       "      <td>798</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Post (prior bilateral ovariectomy OR &gt;12 mo si...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-A2-A04W</td>\n",
       "      <td>Alive</td>\n",
       "      <td>3102</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TCGA-A7-A26G</td>\n",
       "      <td>Alive</td>\n",
       "      <td>722</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>BLACK OR AFRICAN AMERICAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TCGA-AO-A0J5</td>\n",
       "      <td>Dead</td>\n",
       "      <td>792</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>TCGA-AO-A124</td>\n",
       "      <td>Alive</td>\n",
       "      <td>3506</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>TCGA-B6-A0WX</td>\n",
       "      <td>Dead</td>\n",
       "      <td>639</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>BLACK OR AFRICAN AMERICAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>TCGA-C8-A1HL</td>\n",
       "      <td>Alive</td>\n",
       "      <td>317</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>ASIAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>TCGA-C8-A9FZ</td>\n",
       "      <td>Alive</td>\n",
       "      <td>592</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>TCGA-D8-A1X7</td>\n",
       "      <td>Alive</td>\n",
       "      <td>509</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>TCGA-E2-A15A</td>\n",
       "      <td>Alive</td>\n",
       "      <td>710</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>TCGA-E2-A15E</td>\n",
       "      <td>Alive</td>\n",
       "      <td>630</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>TCGA-EW-A2FV</td>\n",
       "      <td>Alive</td>\n",
       "      <td>788</td>\n",
       "      <td>HISPANIC OR LATINO</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>TCGA-A2-A3KD</td>\n",
       "      <td>Alive</td>\n",
       "      <td>1206</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Mucinous Carcinoma</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>TCGA-E9-A5UO</td>\n",
       "      <td>Alive</td>\n",
       "      <td>785</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Mucinous Carcinoma</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>TCGA-EW-A1OX</td>\n",
       "      <td>Alive</td>\n",
       "      <td>911</td>\n",
       "      <td>HISPANIC OR LATINO</td>\n",
       "      <td>Mucinous Carcinoma</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>TCGA-AR-A1AI</td>\n",
       "      <td>Alive</td>\n",
       "      <td>3296</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Medullary Carcinoma</td>\n",
       "      <td>Post (prior bilateral ovariectomy OR &gt;12 mo si...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>TCGA-E9-A3QA</td>\n",
       "      <td>Alive</td>\n",
       "      <td>918</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Medullary Carcinoma</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>TCGA-EW-A1P4</td>\n",
       "      <td>Alive</td>\n",
       "      <td>907</td>\n",
       "      <td>HISPANIC OR LATINO</td>\n",
       "      <td>Medullary Carcinoma</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>TCGA-AC-A2QJ</td>\n",
       "      <td>Dead</td>\n",
       "      <td>446</td>\n",
       "      <td>HISPANIC OR LATINO</td>\n",
       "      <td>Metaplastic Carcinoma</td>\n",
       "      <td>Peri (6-12 months since last menstrual period)</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>TCGA-AR-A2LR</td>\n",
       "      <td>Alive</td>\n",
       "      <td>1742</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Metaplastic Carcinoma</td>\n",
       "      <td>Post (prior bilateral ovariectomy OR &gt;12 mo si...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>TCGA-BH-A18M</td>\n",
       "      <td>Dead</td>\n",
       "      <td>2207</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Infiltrating Ductal Carcinoma</td>\n",
       "      <td>None</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>TCGA-BH-A18V</td>\n",
       "      <td>Dead</td>\n",
       "      <td>1556</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Infiltrating Ductal Carcinoma</td>\n",
       "      <td>None</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>TCGA-A1-A0SJ</td>\n",
       "      <td>Alive</td>\n",
       "      <td>416</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Infiltrating Ductal Carcinoma</td>\n",
       "      <td>None</td>\n",
       "      <td>BLACK OR AFRICAN AMERICAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>TCGA-A1-A0SP</td>\n",
       "      <td>Alive</td>\n",
       "      <td>584</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Infiltrating Ductal Carcinoma</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>TCGA-AC-A23E</td>\n",
       "      <td>Alive</td>\n",
       "      <td>698</td>\n",
       "      <td>None</td>\n",
       "      <td>Infiltrating Ductal Carcinoma</td>\n",
       "      <td>None</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>TCGA-AC-A6NO</td>\n",
       "      <td>Alive</td>\n",
       "      <td>51</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Infiltrating Ductal Carcinoma</td>\n",
       "      <td>None</td>\n",
       "      <td>BLACK OR AFRICAN AMERICAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>TCGA-AC-A2BM</td>\n",
       "      <td>Alive</td>\n",
       "      <td>3022</td>\n",
       "      <td>None</td>\n",
       "      <td>Infiltrating Ductal Carcinoma</td>\n",
       "      <td>None</td>\n",
       "      <td>BLACK OR AFRICAN AMERICAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299</th>\n",
       "      <td>TCGA-A7-A5ZX</td>\n",
       "      <td>Alive</td>\n",
       "      <td>336</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Infiltrating Lobular Carcinoma</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>TCGA-A7-A4SA</td>\n",
       "      <td>Alive</td>\n",
       "      <td>454</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Infiltrating Lobular Carcinoma</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>301</th>\n",
       "      <td>TCGA-A7-A426</td>\n",
       "      <td>Alive</td>\n",
       "      <td>364</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Infiltrating Lobular Carcinoma</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>302</th>\n",
       "      <td>TCGA-AO-A1KO</td>\n",
       "      <td>Alive</td>\n",
       "      <td>622</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Infiltrating Lobular Carcinoma</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>303</th>\n",
       "      <td>TCGA-AR-A2LO</td>\n",
       "      <td>Alive</td>\n",
       "      <td>1198</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Infiltrating Lobular Carcinoma</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304</th>\n",
       "      <td>TCGA-AR-A2LM</td>\n",
       "      <td>Alive</td>\n",
       "      <td>1935</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Infiltrating Lobular Carcinoma</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305</th>\n",
       "      <td>TCGA-AR-A24T</td>\n",
       "      <td>Alive</td>\n",
       "      <td>3202</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Infiltrating Lobular Carcinoma</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>306</th>\n",
       "      <td>TCGA-AR-A2LJ</td>\n",
       "      <td>Alive</td>\n",
       "      <td>2632</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Infiltrating Lobular Carcinoma</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>307</th>\n",
       "      <td>TCGA-BH-A2L8</td>\n",
       "      <td>Alive</td>\n",
       "      <td>612</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Infiltrating Lobular Carcinoma</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>308</th>\n",
       "      <td>TCGA-BH-A0B5</td>\n",
       "      <td>Alive</td>\n",
       "      <td>2136</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Infiltrating Lobular Carcinoma</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>309</th>\n",
       "      <td>TCGA-BH-A28Q</td>\n",
       "      <td>Alive</td>\n",
       "      <td>1119</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Infiltrating Lobular Carcinoma</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>TCGA-E2-A1B5</td>\n",
       "      <td>Alive</td>\n",
       "      <td>984</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Infiltrating Lobular Carcinoma</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311</th>\n",
       "      <td>TCGA-E2-A3DX</td>\n",
       "      <td>Alive</td>\n",
       "      <td>1325</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Infiltrating Lobular Carcinoma</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312</th>\n",
       "      <td>TCGA-E2-A10F</td>\n",
       "      <td>Alive</td>\n",
       "      <td>878</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Infiltrating Lobular Carcinoma</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313</th>\n",
       "      <td>TCGA-E9-A3HO</td>\n",
       "      <td>Alive</td>\n",
       "      <td>1158</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Infiltrating Lobular Carcinoma</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>314</th>\n",
       "      <td>TCGA-E9-A6HE</td>\n",
       "      <td>Alive</td>\n",
       "      <td>847</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Infiltrating Lobular Carcinoma</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315</th>\n",
       "      <td>TCGA-EW-A1J2</td>\n",
       "      <td>Alive</td>\n",
       "      <td>403</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Infiltrating Lobular Carcinoma</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>BLACK OR AFRICAN AMERICAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>316</th>\n",
       "      <td>TCGA-GM-A3XG</td>\n",
       "      <td>Alive</td>\n",
       "      <td>1330</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Infiltrating Lobular Carcinoma</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>317</th>\n",
       "      <td>TCGA-HN-A2OB</td>\n",
       "      <td>Dead</td>\n",
       "      <td>1900</td>\n",
       "      <td>None</td>\n",
       "      <td>Infiltrating Lobular Carcinoma</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>TCGA-OL-A6VQ</td>\n",
       "      <td>Alive</td>\n",
       "      <td>600</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Infiltrating Lobular Carcinoma</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>319</th>\n",
       "      <td>TCGA-PE-A5DE</td>\n",
       "      <td>Alive</td>\n",
       "      <td>2645</td>\n",
       "      <td>None</td>\n",
       "      <td>Infiltrating Lobular Carcinoma</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320</th>\n",
       "      <td>TCGA-B6-A3ZX</td>\n",
       "      <td>Dead</td>\n",
       "      <td>1152</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Mixed Histology (please specify)</td>\n",
       "      <td>None</td>\n",
       "      <td>BLACK OR AFRICAN AMERICAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321</th>\n",
       "      <td>TCGA-B6-A0RV</td>\n",
       "      <td>Alive</td>\n",
       "      <td>5156</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Mixed Histology (please specify)</td>\n",
       "      <td>Indeterminate (neither Pre or Postmenopausal)</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>322</th>\n",
       "      <td>TCGA-E9-A1N5</td>\n",
       "      <td>Alive</td>\n",
       "      <td>1120</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Mixed Histology (please specify)</td>\n",
       "      <td>Post (prior bilateral ovariectomy OR &gt;12 mo si...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323</th>\n",
       "      <td>TCGA-AR-A24O</td>\n",
       "      <td>Alive</td>\n",
       "      <td>3607</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Mixed Histology (please specify)</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>324</th>\n",
       "      <td>TCGA-AR-A24M</td>\n",
       "      <td>Alive</td>\n",
       "      <td>3660</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Mixed Histology (please specify)</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325</th>\n",
       "      <td>TCGA-B6-A0IE</td>\n",
       "      <td>Dead</td>\n",
       "      <td>1993</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Mixed Histology (please specify)</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>BLACK OR AFRICAN AMERICAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>326</th>\n",
       "      <td>TCGA-BH-A0C3</td>\n",
       "      <td>Alive</td>\n",
       "      <td>2709</td>\n",
       "      <td>None</td>\n",
       "      <td>Mixed Histology (please specify)</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>327</th>\n",
       "      <td>TCGA-E9-A1RI</td>\n",
       "      <td>Alive</td>\n",
       "      <td>1449</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Mixed Histology (please specify)</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328</th>\n",
       "      <td>TCGA-E9-A1N4</td>\n",
       "      <td>Alive</td>\n",
       "      <td>1000</td>\n",
       "      <td>NOT HISPANIC OR LATINO</td>\n",
       "      <td>Mixed Histology (please specify)</td>\n",
       "      <td>Pre (&lt;6 months since LMP AND no prior bilatera...</td>\n",
       "      <td>WHITE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>329 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    ParticipantBarcode vital_status  days_to_last_known_alive  \\\n",
       "0         TCGA-B6-A0IN         Dead                      2573   \n",
       "1         TCGA-B6-A0RU        Alive                      8605   \n",
       "2         TCGA-EW-A1IX        Alive                      1208   \n",
       "3         TCGA-AO-A0JG        Alive                       798   \n",
       "4         TCGA-A2-A04W        Alive                      3102   \n",
       "5         TCGA-A7-A26G        Alive                       722   \n",
       "6         TCGA-AO-A0J5         Dead                       792   \n",
       "7         TCGA-AO-A124        Alive                      3506   \n",
       "8         TCGA-B6-A0WX         Dead                       639   \n",
       "9         TCGA-C8-A1HL        Alive                       317   \n",
       "10        TCGA-C8-A9FZ        Alive                       592   \n",
       "11        TCGA-D8-A1X7        Alive                       509   \n",
       "12        TCGA-E2-A15A        Alive                       710   \n",
       "13        TCGA-E2-A15E        Alive                       630   \n",
       "14        TCGA-EW-A2FV        Alive                       788   \n",
       "15        TCGA-A2-A3KD        Alive                      1206   \n",
       "16        TCGA-E9-A5UO        Alive                       785   \n",
       "17        TCGA-EW-A1OX        Alive                       911   \n",
       "18        TCGA-AR-A1AI        Alive                      3296   \n",
       "19        TCGA-E9-A3QA        Alive                       918   \n",
       "20        TCGA-EW-A1P4        Alive                       907   \n",
       "21        TCGA-AC-A2QJ         Dead                       446   \n",
       "22        TCGA-AR-A2LR        Alive                      1742   \n",
       "23        TCGA-BH-A18M         Dead                      2207   \n",
       "24        TCGA-BH-A18V         Dead                      1556   \n",
       "25        TCGA-A1-A0SJ        Alive                       416   \n",
       "26        TCGA-A1-A0SP        Alive                       584   \n",
       "27        TCGA-AC-A23E        Alive                       698   \n",
       "28        TCGA-AC-A6NO        Alive                        51   \n",
       "29        TCGA-AC-A2BM        Alive                      3022   \n",
       "..                 ...          ...                       ...   \n",
       "299       TCGA-A7-A5ZX        Alive                       336   \n",
       "300       TCGA-A7-A4SA        Alive                       454   \n",
       "301       TCGA-A7-A426        Alive                       364   \n",
       "302       TCGA-AO-A1KO        Alive                       622   \n",
       "303       TCGA-AR-A2LO        Alive                      1198   \n",
       "304       TCGA-AR-A2LM        Alive                      1935   \n",
       "305       TCGA-AR-A24T        Alive                      3202   \n",
       "306       TCGA-AR-A2LJ        Alive                      2632   \n",
       "307       TCGA-BH-A2L8        Alive                       612   \n",
       "308       TCGA-BH-A0B5        Alive                      2136   \n",
       "309       TCGA-BH-A28Q        Alive                      1119   \n",
       "310       TCGA-E2-A1B5        Alive                       984   \n",
       "311       TCGA-E2-A3DX        Alive                      1325   \n",
       "312       TCGA-E2-A10F        Alive                       878   \n",
       "313       TCGA-E9-A3HO        Alive                      1158   \n",
       "314       TCGA-E9-A6HE        Alive                       847   \n",
       "315       TCGA-EW-A1J2        Alive                       403   \n",
       "316       TCGA-GM-A3XG        Alive                      1330   \n",
       "317       TCGA-HN-A2OB         Dead                      1900   \n",
       "318       TCGA-OL-A6VQ        Alive                       600   \n",
       "319       TCGA-PE-A5DE        Alive                      2645   \n",
       "320       TCGA-B6-A3ZX         Dead                      1152   \n",
       "321       TCGA-B6-A0RV        Alive                      5156   \n",
       "322       TCGA-E9-A1N5        Alive                      1120   \n",
       "323       TCGA-AR-A24O        Alive                      3607   \n",
       "324       TCGA-AR-A24M        Alive                      3660   \n",
       "325       TCGA-B6-A0IE         Dead                      1993   \n",
       "326       TCGA-BH-A0C3        Alive                      2709   \n",
       "327       TCGA-E9-A1RI        Alive                      1449   \n",
       "328       TCGA-E9-A1N4        Alive                      1000   \n",
       "\n",
       "                  ethnicity                 histological_type  \\\n",
       "0    NOT HISPANIC OR LATINO                    Other, specify   \n",
       "1    NOT HISPANIC OR LATINO                    Other, specify   \n",
       "2        HISPANIC OR LATINO                    Other, specify   \n",
       "3    NOT HISPANIC OR LATINO                    Other, specify   \n",
       "4    NOT HISPANIC OR LATINO                    Other, specify   \n",
       "5    NOT HISPANIC OR LATINO                    Other, specify   \n",
       "6    NOT HISPANIC OR LATINO                    Other, specify   \n",
       "7    NOT HISPANIC OR LATINO                    Other, specify   \n",
       "8    NOT HISPANIC OR LATINO                    Other, specify   \n",
       "9    NOT HISPANIC OR LATINO                    Other, specify   \n",
       "10   NOT HISPANIC OR LATINO                    Other, specify   \n",
       "11   NOT HISPANIC OR LATINO                    Other, specify   \n",
       "12   NOT HISPANIC OR LATINO                    Other, specify   \n",
       "13   NOT HISPANIC OR LATINO                    Other, specify   \n",
       "14       HISPANIC OR LATINO                    Other, specify   \n",
       "15   NOT HISPANIC OR LATINO                Mucinous Carcinoma   \n",
       "16   NOT HISPANIC OR LATINO                Mucinous Carcinoma   \n",
       "17       HISPANIC OR LATINO                Mucinous Carcinoma   \n",
       "18   NOT HISPANIC OR LATINO               Medullary Carcinoma   \n",
       "19   NOT HISPANIC OR LATINO               Medullary Carcinoma   \n",
       "20       HISPANIC OR LATINO               Medullary Carcinoma   \n",
       "21       HISPANIC OR LATINO             Metaplastic Carcinoma   \n",
       "22   NOT HISPANIC OR LATINO             Metaplastic Carcinoma   \n",
       "23   NOT HISPANIC OR LATINO     Infiltrating Ductal Carcinoma   \n",
       "24   NOT HISPANIC OR LATINO     Infiltrating Ductal Carcinoma   \n",
       "25   NOT HISPANIC OR LATINO     Infiltrating Ductal Carcinoma   \n",
       "26   NOT HISPANIC OR LATINO     Infiltrating Ductal Carcinoma   \n",
       "27                     None     Infiltrating Ductal Carcinoma   \n",
       "28   NOT HISPANIC OR LATINO     Infiltrating Ductal Carcinoma   \n",
       "29                     None     Infiltrating Ductal Carcinoma   \n",
       "..                      ...                               ...   \n",
       "299  NOT HISPANIC OR LATINO    Infiltrating Lobular Carcinoma   \n",
       "300  NOT HISPANIC OR LATINO    Infiltrating Lobular Carcinoma   \n",
       "301  NOT HISPANIC OR LATINO    Infiltrating Lobular Carcinoma   \n",
       "302  NOT HISPANIC OR LATINO    Infiltrating Lobular Carcinoma   \n",
       "303  NOT HISPANIC OR LATINO    Infiltrating Lobular Carcinoma   \n",
       "304  NOT HISPANIC OR LATINO    Infiltrating Lobular Carcinoma   \n",
       "305  NOT HISPANIC OR LATINO    Infiltrating Lobular Carcinoma   \n",
       "306  NOT HISPANIC OR LATINO    Infiltrating Lobular Carcinoma   \n",
       "307  NOT HISPANIC OR LATINO    Infiltrating Lobular Carcinoma   \n",
       "308  NOT HISPANIC OR LATINO    Infiltrating Lobular Carcinoma   \n",
       "309  NOT HISPANIC OR LATINO    Infiltrating Lobular Carcinoma   \n",
       "310  NOT HISPANIC OR LATINO    Infiltrating Lobular Carcinoma   \n",
       "311  NOT HISPANIC OR LATINO    Infiltrating Lobular Carcinoma   \n",
       "312  NOT HISPANIC OR LATINO    Infiltrating Lobular Carcinoma   \n",
       "313  NOT HISPANIC OR LATINO    Infiltrating Lobular Carcinoma   \n",
       "314  NOT HISPANIC OR LATINO    Infiltrating Lobular Carcinoma   \n",
       "315  NOT HISPANIC OR LATINO    Infiltrating Lobular Carcinoma   \n",
       "316  NOT HISPANIC OR LATINO    Infiltrating Lobular Carcinoma   \n",
       "317                    None    Infiltrating Lobular Carcinoma   \n",
       "318  NOT HISPANIC OR LATINO    Infiltrating Lobular Carcinoma   \n",
       "319                    None    Infiltrating Lobular Carcinoma   \n",
       "320  NOT HISPANIC OR LATINO  Mixed Histology (please specify)   \n",
       "321  NOT HISPANIC OR LATINO  Mixed Histology (please specify)   \n",
       "322  NOT HISPANIC OR LATINO  Mixed Histology (please specify)   \n",
       "323  NOT HISPANIC OR LATINO  Mixed Histology (please specify)   \n",
       "324  NOT HISPANIC OR LATINO  Mixed Histology (please specify)   \n",
       "325  NOT HISPANIC OR LATINO  Mixed Histology (please specify)   \n",
       "326                    None  Mixed Histology (please specify)   \n",
       "327  NOT HISPANIC OR LATINO  Mixed Histology (please specify)   \n",
       "328  NOT HISPANIC OR LATINO  Mixed Histology (please specify)   \n",
       "\n",
       "                                      menopause_status  \\\n",
       "0        Indeterminate (neither Pre or Postmenopausal)   \n",
       "1        Indeterminate (neither Pre or Postmenopausal)   \n",
       "2       Peri (6-12 months since last menstrual period)   \n",
       "3    Post (prior bilateral ovariectomy OR >12 mo si...   \n",
       "4    Pre (<6 months since LMP AND no prior bilatera...   \n",
       "5    Pre (<6 months since LMP AND no prior bilatera...   \n",
       "6    Pre (<6 months since LMP AND no prior bilatera...   \n",
       "7    Pre (<6 months since LMP AND no prior bilatera...   \n",
       "8    Pre (<6 months since LMP AND no prior bilatera...   \n",
       "9    Pre (<6 months since LMP AND no prior bilatera...   \n",
       "10   Pre (<6 months since LMP AND no prior bilatera...   \n",
       "11   Pre (<6 months since LMP AND no prior bilatera...   \n",
       "12   Pre (<6 months since LMP AND no prior bilatera...   \n",
       "13   Pre (<6 months since LMP AND no prior bilatera...   \n",
       "14   Pre (<6 months since LMP AND no prior bilatera...   \n",
       "15   Pre (<6 months since LMP AND no prior bilatera...   \n",
       "16   Pre (<6 months since LMP AND no prior bilatera...   \n",
       "17   Pre (<6 months since LMP AND no prior bilatera...   \n",
       "18   Post (prior bilateral ovariectomy OR >12 mo si...   \n",
       "19   Pre (<6 months since LMP AND no prior bilatera...   \n",
       "20   Pre (<6 months since LMP AND no prior bilatera...   \n",
       "21      Peri (6-12 months since last menstrual period)   \n",
       "22   Post (prior bilateral ovariectomy OR >12 mo si...   \n",
       "23                                                None   \n",
       "24                                                None   \n",
       "25                                                None   \n",
       "26                                                None   \n",
       "27                                                None   \n",
       "28                                                None   \n",
       "29                                                None   \n",
       "..                                                 ...   \n",
       "299  Pre (<6 months since LMP AND no prior bilatera...   \n",
       "300  Pre (<6 months since LMP AND no prior bilatera...   \n",
       "301  Pre (<6 months since LMP AND no prior bilatera...   \n",
       "302  Pre (<6 months since LMP AND no prior bilatera...   \n",
       "303  Pre (<6 months since LMP AND no prior bilatera...   \n",
       "304  Pre (<6 months since LMP AND no prior bilatera...   \n",
       "305  Pre (<6 months since LMP AND no prior bilatera...   \n",
       "306  Pre (<6 months since LMP AND no prior bilatera...   \n",
       "307  Pre (<6 months since LMP AND no prior bilatera...   \n",
       "308  Pre (<6 months since LMP AND no prior bilatera...   \n",
       "309  Pre (<6 months since LMP AND no prior bilatera...   \n",
       "310  Pre (<6 months since LMP AND no prior bilatera...   \n",
       "311  Pre (<6 months since LMP AND no prior bilatera...   \n",
       "312  Pre (<6 months since LMP AND no prior bilatera...   \n",
       "313  Pre (<6 months since LMP AND no prior bilatera...   \n",
       "314  Pre (<6 months since LMP AND no prior bilatera...   \n",
       "315  Pre (<6 months since LMP AND no prior bilatera...   \n",
       "316  Pre (<6 months since LMP AND no prior bilatera...   \n",
       "317  Pre (<6 months since LMP AND no prior bilatera...   \n",
       "318  Pre (<6 months since LMP AND no prior bilatera...   \n",
       "319  Pre (<6 months since LMP AND no prior bilatera...   \n",
       "320                                               None   \n",
       "321      Indeterminate (neither Pre or Postmenopausal)   \n",
       "322  Post (prior bilateral ovariectomy OR >12 mo si...   \n",
       "323  Pre (<6 months since LMP AND no prior bilatera...   \n",
       "324  Pre (<6 months since LMP AND no prior bilatera...   \n",
       "325  Pre (<6 months since LMP AND no prior bilatera...   \n",
       "326  Pre (<6 months since LMP AND no prior bilatera...   \n",
       "327  Pre (<6 months since LMP AND no prior bilatera...   \n",
       "328  Pre (<6 months since LMP AND no prior bilatera...   \n",
       "\n",
       "                          race  \n",
       "0                        WHITE  \n",
       "1                        WHITE  \n",
       "2                        ASIAN  \n",
       "3                        WHITE  \n",
       "4                        WHITE  \n",
       "5    BLACK OR AFRICAN AMERICAN  \n",
       "6                        WHITE  \n",
       "7                        WHITE  \n",
       "8    BLACK OR AFRICAN AMERICAN  \n",
       "9                        ASIAN  \n",
       "10                       WHITE  \n",
       "11                       WHITE  \n",
       "12                       WHITE  \n",
       "13                       WHITE  \n",
       "14                       WHITE  \n",
       "15                       WHITE  \n",
       "16                       WHITE  \n",
       "17                       WHITE  \n",
       "18                       WHITE  \n",
       "19                       WHITE  \n",
       "20                       WHITE  \n",
       "21                       WHITE  \n",
       "22                       WHITE  \n",
       "23                       WHITE  \n",
       "24                       WHITE  \n",
       "25   BLACK OR AFRICAN AMERICAN  \n",
       "26                        None  \n",
       "27                       WHITE  \n",
       "28   BLACK OR AFRICAN AMERICAN  \n",
       "29   BLACK OR AFRICAN AMERICAN  \n",
       "..                         ...  \n",
       "299                      WHITE  \n",
       "300                      WHITE  \n",
       "301                      WHITE  \n",
       "302                      WHITE  \n",
       "303                      WHITE  \n",
       "304                      WHITE  \n",
       "305                      WHITE  \n",
       "306                      WHITE  \n",
       "307                      WHITE  \n",
       "308                      WHITE  \n",
       "309                      WHITE  \n",
       "310                      WHITE  \n",
       "311                      WHITE  \n",
       "312                      WHITE  \n",
       "313                      WHITE  \n",
       "314                      WHITE  \n",
       "315  BLACK OR AFRICAN AMERICAN  \n",
       "316                      WHITE  \n",
       "317                       None  \n",
       "318                      WHITE  \n",
       "319                      WHITE  \n",
       "320  BLACK OR AFRICAN AMERICAN  \n",
       "321                      WHITE  \n",
       "322                      WHITE  \n",
       "323                      WHITE  \n",
       "324                      WHITE  \n",
       "325  BLACK OR AFRICAN AMERICAN  \n",
       "326                      WHITE  \n",
       "327                      WHITE  \n",
       "328                      WHITE  \n",
       "\n",
       "[329 rows x 7 columns]"
      ]
     },
     "execution_count": 205,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "select_on_clinical.execute().result().to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "and finally we call the main query:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ParticipantBarcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-3C-AALI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-4H-AAAK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-5L-AAT0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-A1-A0SH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-A1-A0SJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TCGA-A1-A0SN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TCGA-A1-A0SP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>TCGA-A1-A0SQ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>TCGA-A2-A04P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>TCGA-A2-A04Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>TCGA-A2-A04R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>TCGA-A2-A04U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>TCGA-A2-A04V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>TCGA-A2-A04W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>TCGA-A2-A04X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>TCGA-A2-A0CL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>TCGA-A2-A0CM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>TCGA-A2-A0CV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>TCGA-A2-A0CZ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>TCGA-A2-A0D2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>TCGA-A2-A0D3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>TCGA-A2-A0D4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>TCGA-A2-A0EX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>TCGA-A2-A0SX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>TCGA-A2-A0T3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>TCGA-A2-A0T5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>TCGA-A2-A0T6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>TCGA-A2-A0YE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>TCGA-A2-A0YJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>TCGA-A2-A0YL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>TCGA-GM-A2DL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>298</th>\n",
       "      <td>TCGA-GM-A3XG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299</th>\n",
       "      <td>TCGA-GM-A3XL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>TCGA-GM-A3XN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>301</th>\n",
       "      <td>TCGA-HN-A2OB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>302</th>\n",
       "      <td>TCGA-JL-A3YW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>303</th>\n",
       "      <td>TCGA-JL-A3YX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304</th>\n",
       "      <td>TCGA-LD-A66U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305</th>\n",
       "      <td>TCGA-LL-A5YN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>306</th>\n",
       "      <td>TCGA-LL-A5YO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>307</th>\n",
       "      <td>TCGA-LL-A5YP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>308</th>\n",
       "      <td>TCGA-LL-A6FR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>309</th>\n",
       "      <td>TCGA-LL-A7SZ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>TCGA-MS-A51U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311</th>\n",
       "      <td>TCGA-OL-A5D8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312</th>\n",
       "      <td>TCGA-OL-A5RV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313</th>\n",
       "      <td>TCGA-OL-A5RW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>314</th>\n",
       "      <td>TCGA-OL-A66I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315</th>\n",
       "      <td>TCGA-OL-A66O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>316</th>\n",
       "      <td>TCGA-OL-A6VO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>317</th>\n",
       "      <td>TCGA-OL-A6VQ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>TCGA-OL-A6VR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>319</th>\n",
       "      <td>TCGA-PE-A5DE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320</th>\n",
       "      <td>TCGA-PL-A8LX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321</th>\n",
       "      <td>TCGA-PL-A8LY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>322</th>\n",
       "      <td>TCGA-PL-A8LZ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323</th>\n",
       "      <td>TCGA-S3-A6ZH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>324</th>\n",
       "      <td>TCGA-S3-AA14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325</th>\n",
       "      <td>TCGA-XX-A899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>326</th>\n",
       "      <td>TCGA-Z7-A8R6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>327 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    ParticipantBarcode\n",
       "0         TCGA-3C-AALI\n",
       "1         TCGA-4H-AAAK\n",
       "2         TCGA-5L-AAT0\n",
       "3         TCGA-A1-A0SH\n",
       "4         TCGA-A1-A0SJ\n",
       "5         TCGA-A1-A0SN\n",
       "6         TCGA-A1-A0SP\n",
       "7         TCGA-A1-A0SQ\n",
       "8         TCGA-A2-A04P\n",
       "9         TCGA-A2-A04Q\n",
       "10        TCGA-A2-A04R\n",
       "11        TCGA-A2-A04U\n",
       "12        TCGA-A2-A04V\n",
       "13        TCGA-A2-A04W\n",
       "14        TCGA-A2-A04X\n",
       "15        TCGA-A2-A0CL\n",
       "16        TCGA-A2-A0CM\n",
       "17        TCGA-A2-A0CV\n",
       "18        TCGA-A2-A0CZ\n",
       "19        TCGA-A2-A0D2\n",
       "20        TCGA-A2-A0D3\n",
       "21        TCGA-A2-A0D4\n",
       "22        TCGA-A2-A0EX\n",
       "23        TCGA-A2-A0SX\n",
       "24        TCGA-A2-A0T3\n",
       "25        TCGA-A2-A0T5\n",
       "26        TCGA-A2-A0T6\n",
       "27        TCGA-A2-A0YE\n",
       "28        TCGA-A2-A0YJ\n",
       "29        TCGA-A2-A0YL\n",
       "..                 ...\n",
       "297       TCGA-GM-A2DL\n",
       "298       TCGA-GM-A3XG\n",
       "299       TCGA-GM-A3XL\n",
       "300       TCGA-GM-A3XN\n",
       "301       TCGA-HN-A2OB\n",
       "302       TCGA-JL-A3YW\n",
       "303       TCGA-JL-A3YX\n",
       "304       TCGA-LD-A66U\n",
       "305       TCGA-LL-A5YN\n",
       "306       TCGA-LL-A5YO\n",
       "307       TCGA-LL-A5YP\n",
       "308       TCGA-LL-A6FR\n",
       "309       TCGA-LL-A7SZ\n",
       "310       TCGA-MS-A51U\n",
       "311       TCGA-OL-A5D8\n",
       "312       TCGA-OL-A5RV\n",
       "313       TCGA-OL-A5RW\n",
       "314       TCGA-OL-A66I\n",
       "315       TCGA-OL-A66O\n",
       "316       TCGA-OL-A6VO\n",
       "317       TCGA-OL-A6VQ\n",
       "318       TCGA-OL-A6VR\n",
       "319       TCGA-PE-A5DE\n",
       "320       TCGA-PL-A8LX\n",
       "321       TCGA-PL-A8LY\n",
       "322       TCGA-PL-A8LZ\n",
       "323       TCGA-S3-A6ZH\n",
       "324       TCGA-S3-AA14\n",
       "325       TCGA-XX-A899\n",
       "326       TCGA-Z7-A8R6\n",
       "\n",
       "[327 rows x 1 columns]"
      ]
     },
     "execution_count": 206,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "create_cohort.execute().result().to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that we didn't need to call each sub-query individually, we could have just called the main query and gotten the same result.  As you can see, two patients that met the clinical select criteria (which returned 329 patients) were excluded from the final result (which returned 327 patients)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Before we leave off, here are a few useful tricks for working with BigQuery in Cloud Datalab:\n",
    "+ if you want to see the raw SQL, you can just build the query and then print it out (this might be useful, for example, in debugging a query -- you can copy and paste the SQL directly into the BigQuery Web UI);\n",
    "+ if you want to see how much data and which tables are going to be touched by this data, you can use the \"dry run\" option.  (Notice the \"cacheHit\" flag -- if you have recently done a particular query, you will not be charged to repeat it since it will have been cached.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre>WITH select_on_annotations AS (\n",
       "  \n",
       "  SELECT\n",
       "    ParticipantBarcode,\n",
       "    annotationCategoryName AS categoryName,\n",
       "    annotationClassification AS classificationName\n",
       "  FROM\n",
       "    `isb-cgc.tcga_201607_beta.Annotations`\n",
       "  WHERE\n",
       "    ( itemTypeName=\"Patient\"\n",
       "      AND (annotationCategoryName=\"History of unacceptable prior treatment related to a prior/other malignancy\"\n",
       "        OR annotationClassification=\"Redaction\" ) )\n",
       "  GROUP BY\n",
       "    ParticipantBarcode,\n",
       "    categoryName,\n",
       "    classificationName\n",
       "),\n",
       "select_on_clinical AS (\n",
       "  \n",
       "  SELECT\n",
       "    ParticipantBarcode,\n",
       "    vital_status,\n",
       "    days_to_last_known_alive,\n",
       "    ethnicity,\n",
       "    histological_type,\n",
       "    menopause_status,\n",
       "    race\n",
       "  FROM\n",
       "    `isb-cgc.tcga_201607_beta.Clinical_data`\n",
       "  WHERE\n",
       "    ( Study=\"BRCA\"\n",
       "      AND age_at_initial_pathologic_diagnosis&lt;=50\n",
       "      AND gender=\"FEMALE\" )\n",
       ")\n",
       "\n",
       "\n",
       "SELECT\n",
       "  ParticipantBarcode\n",
       "FROM (\n",
       "  SELECT\n",
       "    a.categoryName,\n",
       "    a.classificationName,\n",
       "    c.ParticipantBarcode\n",
       "  FROM select_on_annotations AS a\n",
       "  FULL JOIN select_on_clinical AS c\n",
       "  ON\n",
       "    a.ParticipantBarcode = c.ParticipantBarcode\n",
       "  WHERE\n",
       "    a.ParticipantBarcode IS NOT NULL\n",
       "    OR c.ParticipantBarcode IS NOT NULL\n",
       "  ORDER BY\n",
       "    a.classificationName,\n",
       "    a.categoryName,\n",
       "    c.ParticipantBarcode\n",
       ")\n",
       "WHERE\n",
       "  categoryName IS NULL\n",
       "  AND classificationName IS NULL\n",
       "  AND ParticipantBarcode IS NOT NULL\n",
       "ORDER BY\n",
       "  ParticipantBarcode</pre>"
      ],
      "text/plain": [
       "BigQuery Query - \n",
       "SELECT\n",
       "  ParticipantBarcode\n",
       "FROM (\n",
       "  SELECT\n",
       "    a.categoryName,\n",
       "    a.classificationName,\n",
       "    c.ParticipantBarcode\n",
       "  FROM select_on_annotations AS a\n",
       "  FULL JOIN select_on_clinical AS c\n",
       "  ON\n",
       "    a.ParticipantBarcode = c.ParticipantBarcode\n",
       "  WHERE\n",
       "    a.ParticipantBarcode IS NOT NULL\n",
       "    OR c.ParticipantBarcode IS NOT NULL\n",
       "  ORDER BY\n",
       "    a.classificationName,\n",
       "    a.categoryName,\n",
       "    c.ParticipantBarcode\n",
       ")\n",
       "WHERE\n",
       "  categoryName IS NULL\n",
       "  AND classificationName IS NULL\n",
       "  AND ParticipantBarcode IS NOT NULL\n",
       "ORDER BY\n",
       "  ParticipantBarcode"
      ]
     },
     "execution_count": 207,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "create_cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{u'cacheHit': True,\n",
       " u'referencedTables': [{u'datasetId': u'tcga_201607_beta',\n",
       "   u'projectId': u'isb-cgc',\n",
       "   u'tableId': u'Annotations'},\n",
       "  {u'datasetId': u'tcga_201607_beta',\n",
       "   u'projectId': u'isb-cgc',\n",
       "   u'tableId': u'Clinical_data'}],\n",
       " u'schema': {u'fields': [{u'mode': u'NULLABLE',\n",
       "    u'name': u'ParticipantBarcode',\n",
       "    u'type': u'STRING'}]},\n",
       " u'statementType': u'SELECT',\n",
       " u'totalBytesBilled': u'0',\n",
       " u'totalBytesProcessed': u'785007'}"
      ]
     },
     "execution_count": 208,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "create_cohort.dry_run()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
